src/server/mod.rs

```diff
@@ -333,6 +333,27 @@ impl<'a> BlockingRequestAction<'a> for InitializeRequest {
 
         trace!("init: {:?}", init_options);
 
+        // ls_types::ClientCapabilities is a rather awkward object to use internally
+        // (for instance it doesn't Clone). Instead we pick out the bits of it that we
+        // are going to handle into HandledClientCapabilities. The upside of
+        // using this very simple struct is that it can be kept thread safe
+        // without mutex locking it on every request.
+        let capabilities = {
+            let has_snippet_support = params
+            .capabilities
+            .text_document
+            .as_ref()
+            .and_then(|doc| doc.completion.as_ref())
+            .and_then(|comp| comp.completion_item.as_ref())
+            .and_then(|item| item.snippet_support.as_ref())
+            .unwrap_or(&false)
+            .to_owned();
```

Comment:
Could you factor this out into a static method on HandledClientCapabilites please?

Reply:
ok